src/sdk/pynni/nni/compression/torch/utils/sensitivity_analysis.py

```diff
@@ -0,0 +1,210 @@
+            self.sensitivities[name] = {}
+            for sparsity in self.sparsities:
+                # Calculate the actual prune ratio based on the already pruned ratio
+                sparsity = (
+                    1.0 - self.already_pruned[name]) * sparsity + self.already_pruned[name]
+                # TODO In current L1/L2 Filter Pruner, the 'op_types' is still necessary
+                # I think the L1/L2 Pruner should specify the op_types automaticlly
+                # according to the op_names
+                cfg = [{'sparsity': sparsity, 'op_names': [
+                    name], 'op_types': ['Conv2d']}]
+                pruner = self.Pruner(self.model, cfg)
+                pruner.compress()
+                val_acc = self.val_func(*val_args, **val_kwargs)
+                logger.info('Layer: %s Sparsity: %.2f Accuracy: %.4f',
+                            name, sparsity, val_acc)
+
+                self.sensitivities[name][sparsity] = val_acc
+                pruner._unwrap_model()
+                del pruner
+                # if the accuracy drop already reach the 'early_stop'
+                if val_acc + self.early_stop < self.ori_acc:
+                    break
+
+            # reset the weights pruned by the pruner, because the
+            # input sparsities is sorted, so we donnot need to reset
+            # weight of the layer when the sparsity changes, instead,
+            # we only need reset the weight when the pruning layer changes.
+            self.model.load_state_dict(self.ori_state_dict)
+
+        return self.sensitivities
+
+
+    def export(self, filepath):
+        """
+        Export the results of the sensitivity analysis
+        to a csv file. The firstline of the csv file describe the content
+        structure. The first line is constructed by 'layername' and sparsity
+        list. Each line below records the model accuracy when this layer is
+        under different sparsities. Note that, due to the early_stop option,
+        some layers may not have model accuracies under all sparsities, because
+        its accuracy drop has alreay exceeded the threshold set by the user.
+
+        layername, 0.25, 0.5, 0.75
+        conv1, 0.6, 0.55
+        conv2, 0.61, 0.57, 0.56
+
+        Parameters
+        ----------
+        filepath : str
+            Path of the output file
+        """
+        str_sparsities = [str(x) for x in self.sparsities]
+        header = ['layername'] + str_sparsities
+        with open(filepath, 'w') as csvf:
+            csv_w = csv.writer(csvf)
+            csv_w.writerow(header)
+            for layername in self.sensitivities:
+                row = []
+                row.append(layername)
+                for sparsity in sorted(self.sensitivities[layername].keys()):
+                    row.append(self.sensitivities[layername][sparsity])
+                csv_w.writerow(row)
+
+    def update_already_pruned(self, layername, ratio):
+        """
```

Comment:
where is this `update_already_pruned` method used?

Reply:
Yes, both update_already_pruned and load_state_dict are called in the sensitivity_pruner which may be submitted in another pr.

Reply:
OK, got it, thanks